packages/plugins/src/foundry/decode.ts

```diff
@@ -382,3 +382,96 @@ const parseAdditionalContracts = (
     unlabeledAdditionalContracts: unlabeled,
   }
 }
+
+export const makeCreate2DecodedAction = (
+  create2Address: string,
+  initCodeWithArgs: string,
+  configArtifacts: ConfigArtifacts,
+  fullyQualifiedName?: string
+): DecodedAction => {
+  if (fullyQualifiedName) {
+    const coder = ethers.AbiCoder.defaultAbiCoder()
+    const { artifact } = configArtifacts[fullyQualifiedName]
+    const contractName = fullyQualifiedName.split(':')[1]
+    const iface = new ethers.Interface(artifact.abi)
+    const constructorFragment = iface.fragments.find(
+      ConstructorFragment.isFragment
+    )
+
+    let variables: ParsedVariable = {}
+    if (constructorFragment) {
+      const encodedConstructorArgs = getAbiEncodedConstructorArgs(
+        initCodeWithArgs,
+        artifact.bytecode
+      )
+      const constructorArgsResult = coder.decode(
+        constructorFragment.inputs,
+        encodedConstructorArgs
+      )
+      variables = recursivelyConvertResult(
+        constructorFragment.inputs,
+        constructorArgsResult
+      ) as ParsedVariable
+    }
+
+    return {
+      referenceName: contractName,
+      functionName: 'deploy',
+      variables,
+      address: create2Address,
+    }
+  } else {
+    return {
+      referenceName: create2Address,
+      functionName: 'deploy',
+      variables: [],
+      address: create2Address,
+    }
+  }
+}
+
+export const makeFunctionCallDecodedAction = (
+  to: string,
+  data: string,
+  configArtifacts: ConfigArtifacts,
+  dryRunPath: string,
+  fullyQualifiedName?: string
+): DecodedAction => {
+  if (fullyQualifiedName) {
+    const { artifact } = configArtifacts[fullyQualifiedName]
+    const contractName = fullyQualifiedName.split(':')[1]
+    const iface = new ethers.Interface(artifact.abi)
+
+    // Attempt to decode the call. This will return `undefined` if the call cannot be decoded, which
+    // will happen if the function does not exist on the target contract. For example, this will
+    // return `undefined` if the contract's `fallback` function was called.
+    const decoded = decodeCall(iface, data)
+    const functionName = decoded ? decoded.functionName : 'call'
+    const variables = decoded
+      ? decoded.variables
+      : [
+          data.length > 1000
+            ? `Very large calldata. View it in Foundry's dry run file: ${dryRunPath}`
+            : data,
+        ]
+
+    return {
+      referenceName: contractName,
+      functionName,
+      variables,
+      address: to,
+    }
+  } else {
+    const variables = [
+      data.length > 1000
+        ? `Very large calldata. View it in Foundry's dry run file: ${dryRunPath}`
+        : data,
+    ]
+    return {
+      referenceName: to,
+      functionName: 'call',
+      variables,
+      address: '',
+    }
```

Comment:
It's rather confusing that function call actions with no fully qualified name have a `referenceName` field equal to the `to` address, and an `address` field equal to an empty string. However, this is consistent with the previous implementation for the decoded actions, and changing this is out of scope for this PR

Reply:
Yeah, I agree this is weird. It was just implemented this way because it didn't require refactoring the rendering function. I've created a ticket for it.

Reply:
There's an [existing ticket](https://linear.app/chugsplash/issue/CHU-324/improve-preview) for various preview improvements. I think it's already in there